File: code/programs/typescript/unix-tools/tests/whoami.test.ts
/**
 * Tests for whoami -- print effective user name.
 *
 * We test the exported `getEffectiveUsername` function, which retrieves
 * the current user's name via `os.userInfo()` with a fallback to
 * `process.env.USER`.
 *
 * Note: ESM namespaces cannot be spied on with vi.spyOn, so we cannot
 * mock `os.userInfo()`. Instead, we test the function's behavior with
 * the real system state and exercise the fallback path by manipulating
 * environment variables (which IS allowed).
 */

import { describe, it, expect, afterEach } from "vitest";
import * as os from "node:os";
import { getEffectiveUsername } from "../src/whoami.js";

describe("getEffectiveUsername", () => {
  // -------------------------------------------------------------------------
  // Normal operation
  // -------------------------------------------------------------------------

  it("should return a non-empty string", () => {
    const username = getEffectiveUsername();
    expect(username).toBeTruthy();
    expect(typeof username).toBe("string");
  });

  it("should return the same value as os.userInfo().username", () => {
    const expected = os.userInfo().username;
    const result = getEffectiveUsername();
    expect(result).toBe(expected);
  });

  it("should return a string without leading/trailing whitespace", () => {
    const username = getEffectiveUsername();
    expect(username).not.toBeNull();
    expect(username!.trim()).toBe(username);
  });

  it("should return a non-empty username on any normal system", () => {
    const username = getEffectiveUsername();
    expect(username).not.toBeNull();
    expect(username!.length).toBeGreaterThan(0);
  });

  // -------------------------------------------------------------------------
  // Verify it matches the system username
  // -------------------------------------------------------------------------

  it("should match process.env.USER on a normal system", () => {
    // On most systems, os.userInfo().username and process.env.USER agree.
    // This test documents that expectation.
    const username = getEffectiveUsername();
    const envUser = process.env.USER;
    if (envUser) {
      expect(username).toBe(envUser);
    }
  });

  it("should return a string that contains only valid username characters", () => {
    const username = getEffectiveUsername();
    expect(username).not.toBeNull();
    // Unix usernames typically contain only alphanumeric, underscore, hyphen, dot
    expect(username).toMatch(/^[a-zA-Z0-9._-]+$/);
  });
});

File: code/programs/rust/unix-tools/tests/whoami_test.rs
//! # Integration Tests for whoami
//!
//! These tests verify that the `whoami` JSON spec integrates correctly
//! with CLI Builder, and that the business logic returns a valid
//! username from the environment.

use cli_builder::{load_spec_from_file, Parser, ParserOutput};
use std::path::PathBuf;
use unix_tools::whoami_tool::get_username;

// ---------------------------------------------------------------------------
// Helper: locate the spec file
// ---------------------------------------------------------------------------

fn spec_path() -> String {
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let path = PathBuf::from(manifest_dir).join("whoami.json");
    path.to_string_lossy().into_owned()
}

fn make_parser() -> Parser {
    let spec = load_spec_from_file(&spec_path()).expect("failed to load whoami.json");
    Parser::new(spec)
}

fn parse_argv(argv: &[&str]) -> ParserOutput {
    let parser = make_parser();
    let args: Vec<String> = argv.iter().map(|s| s.to_string()).collect();
    parser.parse(&args).expect("parse failed")
}

// ---------------------------------------------------------------------------
// Test: Spec loads successfully
// ---------------------------------------------------------------------------

#[cfg(test)]
mod spec_loading {
    use super::*;

    #[test]
    fn spec_loads() {
        let spec = load_spec_from_file(&spec_path());
        assert!(spec.is_ok(), "whoami.json should load successfully");
    }
}

// ---------------------------------------------------------------------------
// Test: Default behavior
// ---------------------------------------------------------------------------

#[cfg(test)]
mod default_behavior {
    use super::*;

    #[test]
    fn no_flags_returns_parse_result() {
        match parse_argv(&["whoami"]) {
            ParserOutput::Parse(_) => {}
            other => panic!("expected Parse, got {:?}", other),
        }
    }
}

// ---------------------------------------------------------------------------
// Test: --help and --version
// ---------------------------------------------------------------------------

#[cfg(test)]
mod builtins {
    use super::*;

    #[test]
    fn help_returns_help_result() {
        match parse_argv(&["whoami", "--help"]) {
            ParserOutput::Help(_) => {}
            other => panic!("expected Help, got {:?}", other),
        }
    }

    #[test]
    fn help_text_contains_program_name() {
        match parse_argv(&["whoami", "--help"]) {
            ParserOutput::Help(help) => {
                assert!(
                    help.text.contains("whoami"),
                    "help text should contain 'whoami'"
                );
            }
            _ => panic!("expected Help"),
        }
    }

    #[test]
    fn version_returns_version_result() {
        match parse_argv(&["whoami", "--version"]) {
            ParserOutput::Version(_) => {}
            other => panic!("expected Version, got {:?}", other),
        }
    }

    #[test]
    fn version_string() {
        match parse_argv(&["whoami", "--version"]) {
            ParserOutput::Version(v) => {
                assert_eq!(v.version, "1.0.0");
            }
            _ => panic!("expected Version"),
        }
    }
}

// ---------------------------------------------------------------------------
// Test: Business logic — get_username
// ---------------------------------------------------------------------------

#[cfg(test)]
mod business_logic {
    use super::*;

    // These tests rely on $USER/$LOGNAME env vars which are Unix-specific.
    // Windows uses $USERNAME instead, and neither $USER nor $LOGNAME is
    // typically set on Windows CI.

    #[cfg(unix)]
    #[test]
    fn returns_a_username() {
        let result = get_username();
        assert!(result.is_ok(), "get_username should succeed");
        let name = result.unwrap();
        assert!(!name.is_empty(), "username should not be empty");
    }

    #[cfg(unix)]
    #[test]
    fn username_has_no_newlines() {
        let name = get_username().unwrap();
        if name.contains('\n') {
            panic!("username should not contain newlines");
        }
    }

    #[cfg(unix)]
    #[test]
    fn username_is_consistent() {
        let first = get_username().unwrap();
        let second = get_username().unwrap();
        if first != second {
            panic!("username should be consistent across calls");
        }
    }

    #[cfg(unix)]
    #[test]
    fn username_matches_env() {
        // The username should match $USER or $LOGNAME.
        let name = get_username().unwrap();
        let user_var = std::env::var("USER").ok();
        let logname_var = std::env::var("LOGNAME").ok();
        if user_var.as_deref() != Some(name.as_str())
            && logname_var.as_deref() != Some(name.as_str())
        {
            panic!("username should match $USER or $LOGNAME");
        }
    }
}
